File: 202410/1331.py
class Solution:
    def arrayRankTransform(self, arr: List[int]) -> List[int]:
        sa = sorted(arr)
        n = len(arr)
        if n == 0:
            return []
        rank = {}
        rank[sa[0]] = 1
        for i in range(1, n):
            if sa[i] == sa[i - 1]:
                rank[sa[i]] = rank[sa[i - 1]]
            else:
                rank[sa[i]] = rank[sa[i - 1]] + 1

        ans = [0] * n
        for i in range(n):
            ans[i] = rank[arr[i]]
        return ans

File: 202607/1331.cpp
class Solution {
public:
  vector<int> arrayRankTransform(vector<int> &arr) {
    vector<int> sorted(arr);
    sort(begin(sorted), end(sorted));
    sorted.erase(unique(begin(sorted), end(sorted)), end(sorted));
    vector<int> ans(arr.size());
    for (auto [i, x] : std::views::enumerate(arr)) {
      ans[i] = lower_bound(begin(sorted), end(sorted), x) - sorted.begin() + 1;
    }
    return ans;
  }
};
